src/main/scala/sigmastate/lang/SigmaTyper.scala

```diff
@@ -145,18 +145,35 @@ class SigmaTyper {
           if (actualTypes != argTypes)
             error(s"Invalid argument type of application $app: expected $argTypes; actual: $actualTypes")
           Apply(new_f, newArgs)
-        case _: SCollection[_] =>
+        case SCollection(elemType) =>
           // If it's a collection then the application has type of that collection's element.
           args match {
-            case Seq(Constant(i, tpe: SNumericType)) =>
-              ByIndex[SType](new_f.asCollection, SInt.upcast(i.asInstanceOf[AnyVal]))
-            case Seq(arg) =>
-              val newArg = assignType(env, arg)
-              newArg.tpe match {
-                case targ: SNumericType =>
-                  ByIndex[SType](new_f.asCollection, newArg.upcastTo(SInt))
+            case Seq(Constant(index, _: SNumericType)) =>
+              ByIndex[SType](new_f.asCollection, SInt.upcast(index.asInstanceOf[AnyVal]), None)
+            case Seq(Constant(index, _: SNumericType), dvConst@Constant(_, dvtpe)) =>
```

Comment:
@greenhat I cannot accept this syntax, I only agreed with IR part of #127 
Please remove all this special cases in SigmaTyper, they are required because the syntax is ad-hoc.
Please add method `applyOrElse` next to  `"size" -> SInt` in `object SCollection`.
Check where SProduct.fields is used.
I expect no changes in SigmaTyper are necessary to add `applyOrElse`, but you will have to add a rule in SigmaSpecializer.
See for example `case Apply(Select(col,"exists", _),...)`.

Reply:
@ergomorphic Agreed. I find `getOrElse` more appealing, especially for developers with less experience it better conveys the notion that you are getting something out of the collection.